src/windows.rs

```diff
@@ -2,47 +2,38 @@
 
 use std::env;
 use std::ffi::OsString;
-use std::io;
+use std::mem::MaybeUninit;
 use std::os::windows::ffi::OsStringExt;
 use std::path::PathBuf;
 use std::ptr;
+use std::slice;
 
-use winapi::shared::minwindef::DWORD;
-use winapi::shared::winerror::ERROR_INSUFFICIENT_BUFFER;
-use winapi::um::errhandlingapi::{GetLastError, SetLastError};
-use winapi::um::handleapi::CloseHandle;
-use winapi::um::processthreadsapi::{GetCurrentProcess, OpenProcessToken};
-use winapi::um::userenv::GetUserProfileDirectoryW;
-use winapi::um::winnt::TOKEN_READ;
+use winapi::shared::minwindef::MAX_PATH;
+use winapi::shared::winerror::S_OK;
+use winapi::um::shlobj::{SHGetFolderPathW, CSIDL_PROFILE};
 
 pub fn home_dir_inner() -> Option<PathBuf> {
     env::var_os("USERPROFILE")
+        .filter(|s| !s.is_empty())
         .map(PathBuf::from)
         .or_else(home_dir_crt)
 }
 
 #[cfg(not(target_vendor = "uwp"))]
 fn home_dir_crt() -> Option<PathBuf> {
     unsafe {
-        let me = GetCurrentProcess();
-        let mut token = ptr::null_mut();
-        if OpenProcessToken(me, TOKEN_READ, &mut token) == 0 {
-            return None;
+        let mut path: [MaybeUninit<u16>; MAX_PATH] = MaybeUninit::uninit().assume_init();
+        let ptr = path.as_mut_ptr() as *mut u16;
+        match SHGetFolderPathW(ptr::null_mut(), CSIDL_PROFILE, ptr::null_mut(), 0, ptr) {
+            S_OK => {
+                let ptr = path.as_ptr() as *const u16;
+                let len = wcslen(ptr);
+                let path = slice::from_raw_parts(ptr, len);
+                let s = OsString::from_wide(path);
+                Some(PathBuf::from(s))
```

Comment:
Did I do something that expose UBs? cc @retep998 
I know you are busy but this crate is quite important to cargo and rustup.
I would be nice if I don't do something wrong here.